cilium: Reinitialize datapath before annotating node object

jrfastab opened the pull request:
When rolling cilium-agent or doing an upgrade while running stress test
with encryption a small number of NoStateIn errors are seen. To capture
the error state (a cilium_host IP without an xfrm state rule) you need
to get into the pod near pod init and get somewhat lucky that init
took some longer time. For example I ran `ip x s` in a pod about
15seconds after launch and captured a case with new XfrmInNoErrors,
a cilium_host ip assigned, but no xfrm state rule for it. The packets
received are dropped.

The conclusion is remote nodes learn the new router IP before we have
the xfrm state rule loaded. The remote nodes then start using that
IP for the IPSec tunnel outer IP resulting in the errors when they
reach the local node without the xfrm rule yet. The errors eventually
resolve, but some packets are lost in the meantime.

The reason this happens is because first we configure the datapath
after we push node object updates. This is wrong because we need
to init the ipsec code path before we teach remote nodes about the
new IP. And second the configuration of the datapath does a lookup
in the node objects IPAddresses{} this is only populated from the
k8s watcher in the tunnel case. So we only have the fully populated
node object after we receive it through the k8s watcher. Again its
possible other nodes already have seen the event and started pushing
traffic with the new IPs.

To resolve push IPSec init code to create xfrm rules needed with
the new IPs before we publish them to the k8s node object. And
instead of pulling the IPs out of the node object simply pull them
directly from the node module. This resolves the XfrmInNoState and
XfrmIn*Policy* errors I've seen.

To reproduce the errors I can consistently reproduce with about
30 nodes, with httpperf test running from a pod in all nodes, and
then doing a 'rollout' of the cilium agent for awhile. Seems
a 2-3 hours almost ensures errors pop up. Usually the errors
happen much sooner. Initially I saw these errors on upgrade tests
which is another method to reproduce.

pchaigno reviewed (changes requested):
I've read the code and commit description multiple times, but I am unable to understand any of this :disappointed: 

What is the impact of this change? I am guessing connectivity drops; how does the code change described here cause those drops?
What is the relation between node annotations and IPsec?

> Interesting side effect is to know if all subsystems can have two configs in place at the same time.

What kind of config is this referring to? XFRM config?

> IPSec is OK to manage the two states which appears to be requirement to not drop traffic on update while multiple IPs are in the network.

Does that mean we already have logic in place to support having the two "configs" in place at the same time?

jrfastab committed:
cilium: Ensure xfrm state is initialized for route IP before publish

When rolling cilium-agent or doing an upgrade while running stress test
with encryption a small number of NoStateIn errors are seen. To capture
the error state (a cilium_host IP without an xfrm state rule) you need
to get into the pod near pod init and get somewhat lucky that init
took some longer time. For example I ran `ip x s` in a pod about
15seconds after launch and captured a case with new XfrmInNoErrors,
a cilium_host ip assigned, but no xfrm state rule for it. The packets
received are dropped.

The conclusion is remote nodes learn the new router IP before we have
the xfrm state rule loaded. The remote nodes then start using that
IP for the IPSec tunnel outer IP resulting in the errors when they
reach the local node without the xfrm rule yet. The errors eventually
resolve, but some packets are lost in the meantime.

The reason this happens is because first we configure the datapath
after we push node object updates. This is wrong because we need
to init the ipsec code path before we teach remote nodes about the
new IP. And second the configuration of the datapath does a lookup
in the node objects IPAddresses{} this is only populated from the
k8s watcher in the tunnel case. So we only have the fully populated
node object after we receive it through the k8s watcher. Again its
possible other nodes already have seen the event and started pushing
traffic with the new IPs.

To resolve push IPSec init code to create xfrm rules needed with
the new IPs before we publish them to the k8s node object. And
instead of pulling the IPs out of the node object simply pull them
directly from the node module. This resolves the XfrmInNoState and
XfrmIn*Policy* errors I've seen.

To reproduce the errors I can consistently reproduce with about
30 nodes, with httpperf test running from a pod in all nodes, and
then doing a 'rollout' of the cilium agent for awhile. Seems
a 2-3 hours almost ensures errors pop up. Usually the errors
happen much sooner. Initially I saw these errors on upgrade tests
which is another method to reproduce.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -1121,3 +1121,32 @@ func StartStaleKeysReclaimer(ctx context.Context) {
+		if !os.IsExist(err) {
+			return fmt.Errorf("unable to replace policy fwd: %s", err)
+		}
+	}
```
There's a lot of code duplication between this here and `enableIPsecIPv4`/`UpsertIPsecEndpoint`. These XFRM configs also don't exactly match those installed by `enableIPsecIPv4` in that the latter uses `n.nodeConfig.IPv4PodSubnets` as the destination CIDRs for XFRM IN policies. Finally, the IPv6 counterpart is missing.

Maybe to fix all of the above we could move the `newNode.IsLocal()` case of `enableIPsecIPv{4,6}` to this `Init` function instead of duplicating?

jrfastab replied:
One way to refactor would be to make it this?
```
func Init() error {
        outerLocalIP := node.GetInternalIPv4Router()                                                                                                                                   
        wildcardIP := net.ParseIP("0.0.0.0")                                                                                                                                                                                           
        localCIDR := node.GetIPv4AllocRange().IPNet                                                                                                                
        localWildcardIP := &net.IPNet{IP: wildcardIP, Mask: net.IPv4Mask(0, 0, 0, 0)}                                                                      

         _, err := UpsertIPsecEndpoint(localCIDR, localWildcardIP, outerLocalIP, wildcardIP, 0, IPDitIn, gsldr)
        return err;
}
```
I believe that is equivalent would probably want to double check. I prefer that to reaching into pkg/datapath/linux/node.go and start refactoring node code. I think it should probably be done but not in the context of a fix.

pchaigno reviewed (changes requested):
Thanks for the explanation! The commit description also looks a lot better :slightly_smiling_face: 

I think we could however avoid the code duplication with the existing XFRM IN config installations. The two code snippets should do the same thing, but they have already started diverging.
